test/Concurrency/Runtime/async_task_sleep_cancel.swift

```diff
@@ -0,0 +1,70 @@
+// RUN: %target-run-simple-swift(-Xfrontend -enable-experimental-concurrency %import-libdispatch -parse-as-library) | %FileCheck %s --dump-input always
+// REQUIRES: executable_test
+// REQUIRES: concurrency
+// REQUIRES: libdispatch
+
+// rdar://76038845
+// UNSUPPORTED: use_os_stdlib
+// UNSUPPORTED: back_deployment_runtime
+
+import _Concurrency
+// FIXME: should not depend on Dispatch
+import Dispatch
+
+@available(SwiftStdlib 5.5, *)
+@main struct Main {
+  static let pause = 500_000_000 // 500ms
+
+  static func main() async {
+    // CHECK: Starting!
+    print("Starting!")
+    await testSleepFinished()
+    await testSleepCancelled()
+  }
+
+  static func testSleepFinished() async {
+    // CHECK-NEXT: Testing sleep that completes
+    print("Testing sleep that completes")
+    let start = DispatchTime.now()
+
+    // try! will fail if the task got cancelled (which shouldn't happen).
+    try! await Task.sleep(nanoseconds: UInt64(pause))
+
+    let stop = DispatchTime.now()
+
+    // assert that at least the specified time passed since calling `sleep`
+    assert(stop >= (start + .nanoseconds(pause)))
+
+    // CHECK-NEXT: Wakey wakey!
+    print("Wakey wakey!")
+  }
+
+  static func testSleepCancelled() async {
+    // CHECK-NEXT: Testing sleep that gets cancelled before it completes
+    print("Testing sleep that gets cancelled before it completes")
+    let start = DispatchTime.now()
+
+    let sleepyTask = Task {
+      try await Task.sleep(nanoseconds: UInt64(pause))
+    }
+    sleepyTask.cancel()
```

Comment:
I think this also needs a test for if the task is cancelled after it has started.

Reply:
Right, we need both.